enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,210 @@
+
+* As a MicroShift administrator, I want to provide additional DNS names for each certificate.
+
+### Goals
+* Allow MicroShift administrator to configure Microshift with externally generated certificates and
+domain names.
+
+### Non-Goals
+N/A
+
+## Proposal
+Proposal suggests to provide Administrator means adding their own Certificates using microshift configuration file.
+
+A new `apiServer` level section will be added to the configuration called `namedCertificates`:
+
+```yaml
+apiServer:
+  namedCertificates:
+  - certPath: ~/certs/api_fqdn_1.crt
+    keyPath:  ~/certs/api_fqdn_1.key
+  - certPath: ~/certs/api_fqdn_2.crt
+    keyPath:  ~/certs/api_fqdn_2.key
+    names:
+    - api_fqdn_1     
+    - *.apps.external.com
+
+```  
+For each provided certificate, the following configuration is proposed:
+1. `names` - optional list of explicit DNS names (leading wildcards allowed).
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+Certificate files will be read from their configured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+because we dont want to disrupt the internal API communication and make sure the internal certificate will be automaticly renewed,
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+The generated `kubeconfig` for the custom certificates will omit the certificate-authority-data section,
+therefore custom certificates will have to be validated against CAs in the RHEL Client trust store. 
+
+For every non-wildcard certificate CN content will be placed inside the `server` section of the generated `kubeconfig`.
+if the CN Section of the certificate will contain wildcard then `kubeconfig` wont be generated.
+
+Certain values that configured in `names` field can cause  internal API communication issues  ie:127.0.0.1,localhost we must not allow them.
+
+### Workflow Description
+
+#### Default API Certs
+1. By default, when there is no namedCertificates configuration the behaviour will remain the same.
+
+#### when custom namedCertificates configured
+1. Device Administrator copies the certificated to MicroShift host
+1. Device Administrator configures additonal CAs in the RHEL Client trust store.
+1. Device Administrator configures `namedCertificates` in the Microshift configuration yaml file (/etc/microshift/config.yaml).
+1. Device Administrator start/restarts MicroShift
+1. During startup the certificates paths will be checked and validated. in case of an error service will produce clear error log and if the file is missing it will exit.
+1. Certificates will be passed to the tls-sni-cert-key as apiserver command line option preceding all the other certificates.
```

Comment:
Active voice again

```suggestion
1. Microsoft passes the certificates to the tls-sni-cert-key as apiserver command line option preceding all the other certificates.
```

This says specifically "preceding", does the order matter? Maybe it sets the lookup precedence?

Reply:
yes the order matters, since we want the custom certs to be served first on the same SNI (since we also  may generate external cert for  that SNI)